src/app/InteractionModelEngine.h

```diff
@@ -213,9 +219,15 @@ class InteractionModelEngine : public Messaging::ExchangeDelegate, public Comman
     CHIP_ERROR OnUnsolicitedReportData(Messaging::ExchangeContext * apExchangeContext, const PayloadHeader & aPayloadHeader,
                                        System::PacketBufferHandle && aPayload);
 
+    void DispatchCommand(CommandHandler & apCommandObj, const ConcreteCommandPath & aCommandPath,
+                         TLV::TLVReader & apPayload) override;
+    bool CommandExists(const ConcreteCommandPath & aCommandPath) override;
+
     Messaging::ExchangeManager * mpExchangeMgr = nullptr;
     InteractionModelDelegate * mpDelegate      = nullptr;
 
+    CommandHandlerInterface * mpCommandHandlerList = nullptr;
```

Comment:
This list needs to be fully unregistered when the IM engine is shut down.  Otherwise shutdown+reinit could leave us in a place where handlers that were already registered try to be registered again...

Reply:
I'm not sure I follow - the act of unregistering all handlers (if we were to do such a thing) is purely a data structure clean-up, there is no callback to the application to indicate so.

Shutdown+Reinit will very much result in handlers being registered again, which is ... expected, and should work.

Reply:
Here's the problem.  Let's say there is only one handler to be registered, and it lives at address X.

When the stack starts up the first time, `mCommandHandlerList == nullptr`.  The handler gets registered (triggered by what?); now `mCommandHandlerList == X` and the handler's `mNext` is nullptr.

Now the stack is shut down and restarted.  This does not do anything with the handler or `mCommandHandlerList`, right?  Will stack restart trigger the handler registration again?  If it does, then we will end up setting its `mNext` to X, and will have a loop in our linked list.

Reply:
> (triggered by what?)

Could be anything. Most examples revolve around the creation of a new dynamic endpoint that houses a number of server clusters for which the application wants to implement command handlers for.

> This does not do anything with the handler or mCommandHandlerList, right?

Ah yes, I forget that the InteractionModelEngine instance doesn't get re-constructed since it's a static. In that case yes, we should either null it out on the Shutdown, or null it out on Init (likely the latter).

Reply:
That seems like the least we should do.  It still leaves handlers pointing to each other even though they are conceptually not in the list anymore, but that's not observable as long as all the interactions are via the IM engine...  If we wanted to be safer, we'd null out all the "next" pointers of whatever is still registered at shutdown, though.

Reply:
Done.